File: detection/rules.go
package detection

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Rule represents a parsed Snort-like detection rule.
type Rule struct {
	ID       int
	Action   string // alert, log, pass, drop
	Protocol string // tcp, udp, icmp, any
	SrcIP    string
	SrcPort  string
	DstIP    string
	DstPort  string
	Msg      string
	Flags    string // TCP flags: S, A, R, F, P
	Content  string // payload substring match
	Nocase   bool
	Sid      int
	Rev      int
	Severity int // 1=low, 2=medium, 3=high
}

// LoadRules parses a Snort-like rules file.
// Format: action proto src_ip src_port -> dst_ip dst_port (options)
func LoadRules(path string) ([]Rule, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open rules file: %w", err)
	}
	defer f.Close()

	var rules []Rule
	scanner := bufio.NewScanner(f)
	lineNo := 0

	for scanner.Scan() {
		lineNo++
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		r, err := parseRule(line)
		if err != nil {
			return nil, fmt.Errorf("line %d: %w", lineNo, err)
		}
		r.ID = lineNo
		rules = append(rules, r)
	}
	return rules, scanner.Err()
}

func parseRule(line string) (Rule, error) {
	var r Rule

	// Split header from options: everything before first '('
	parenIdx := strings.Index(line, "(")
	if parenIdx < 0 {
		return r, fmt.Errorf("missing options section")
	}

	header := strings.Fields(strings.TrimSpace(line[:parenIdx]))
	if len(header) < 7 {
		return r, fmt.Errorf("short header: %q", line[:parenIdx])
	}

	r.Action = header[0]
	r.Protocol = header[1]
	r.SrcIP = header[2]
	r.SrcPort = header[3]
	// header[4] is the direction indicator "->" or "<>"
	r.DstIP = header[5]
	r.DstPort = header[6]

	// Parse options inside parentheses
	optStr := line[parenIdx+1:]
	if idx := strings.LastIndex(optStr, ")"); idx >= 0 {
		optStr = optStr[:idx]
	}

	for _, opt := range splitOptions(optStr) {
		kv := strings.SplitN(opt, ":", 2)
		key := strings.TrimSpace(kv[0])
		val := ""
		if len(kv) == 2 {
			val = strings.Trim(strings.TrimSpace(kv[1]), "\"")
		}

		switch key {
		case "msg":
			r.Msg = val
		case "flags":
			r.Flags = val
		case "content":
			r.Content = val
		case "nocase":
			r.Nocase = true
		case "sid":
			r.Sid, _ = strconv.Atoi(val)
		case "rev":
			r.Rev, _ = strconv.Atoi(val)
		case "severity":
			r.Severity, _ = strconv.Atoi(val)
		}
	}

	return r, nil
}

// splitOptions splits "msg:\"x\"; flags:S; ..." respecting quoted strings.
func splitOptions(s string) []string {
	var opts []string
	var cur strings.Builder
	inQuote := false

	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '"' {
			inQuote = !inQuote
			cur.WriteByte(c)
		} else if c == ';' && !inQuote {
			opt := strings.TrimSpace(cur.String())
			if opt != "" {
				opts = append(opts, opt)
			}
			cur.Reset()
		} else {
			cur.WriteByte(c)
		}
	}
	if opt := strings.TrimSpace(cur.String()); opt != "" {
		opts = append(opts, opt)
	}
	return opts
}

File: detection/rules.py
from __future__ import annotations
import re
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class Rule:
    id: int = 0
    action: str = "alert"
    protocol: str = "any"
    src_ip: str = "any"
    src_port: str = "any"
    dst_ip: str = "any"
    dst_port: str = "any"
    msg: str = ""
    flags: str = ""
    content: str = ""
    nocase: bool = False
    sid: int = 0
    rev: int = 1
    severity: int = 1


def load_rules(path: str) -> List[Rule]:
    rules: List[Rule] = []
    try:
        with open(path) as f:
            lines = f.readlines()
    except FileNotFoundError:
        return rules

    for lineno, line in enumerate(lines, 1):
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        try:
            rules.append(_parse_rule(line, lineno))
        except ValueError as e:
            raise ValueError(f"line {lineno}: {e}") from e

    return rules


def _parse_rule(line: str, lineno: int) -> Rule:
    paren_idx = line.find("(")
    if paren_idx < 0:
        raise ValueError("missing options section")

    header = line[:paren_idx].split()
    if len(header) < 7:
        raise ValueError(f"short header: {line[:paren_idx]!r}")

    r = Rule(
        id=lineno,
        action=header[0],
        protocol=header[1],
        src_ip=header[2],
        src_port=header[3],
        # header[4] is direction "->"
        dst_ip=header[5],
        dst_port=header[6],
    )

    opt_str = line[paren_idx + 1:]
    last_paren = opt_str.rfind(")")
    if last_paren >= 0:
        opt_str = opt_str[:last_paren]

    for opt in _split_options(opt_str):
        kv = opt.split(":", 1)
        key = kv[0].strip()
        val = kv[1].strip().strip('"') if len(kv) == 2 else ""

        if key == "msg":
            r.msg = val
        elif key == "flags":
            r.flags = val
        elif key == "content":
            r.content = val
        elif key == "nocase":
            r.nocase = True
        elif key == "sid":
            r.sid = int(val)
        elif key == "rev":
            r.rev = int(val)
        elif key == "severity":
            r.severity = int(val)

    return r


def _split_options(s: str) -> List[str]:
    opts: List[str] = []
    current: List[str] = []
    in_quote = False

    for ch in s:
        if ch == '"':
            in_quote = not in_quote
            current.append(ch)
        elif ch == ";" and not in_quote:
            opt = "".join(current).strip()
            if opt:
                opts.append(opt)
            current = []
        else:
            current.append(ch)

    if current:
        opt = "".join(current).strip()
        if opt:
            opts.append(opt)

    return opts
